include/sys/async_notify.h

```diff
@@ -0,0 +1,346 @@
+/*
+ * Copyright (c) 2019 Peter Bigot Consulting, LLC
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_ASYNC_NOTIFY_H_
+#define ZEPHYR_INCLUDE_SYS_ASYNC_NOTIFY_H_
+
+#include <kernel.h>
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup async_notify_apis Asynchronous Notification APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/* Forward declaration */
+struct async_notify;
+
+/*
+ * Flag value that overwrites the method field when the operation has
+ * completed.
+ */
+#define ASYNC_NOTIFY_METHOD_COMPLETED 0
+
+/*
+ * Indicates that no notification will be provided.
+ *
+ * Callers must check for completions using
+ * async_notify_fetch_result().
+ *
+ * See async_notify_init_spinwait().
+ */
+#define ASYNC_NOTIFY_METHOD_SPINWAIT 1
+
+/*
+ * Select notification through @ref k_poll signal
+ *
+ * See async_notify_init_signal().
+ */
+#define ASYNC_NOTIFY_METHOD_SIGNAL 2
+
+/*
+ * Select notification through a user-provided callback.
+ *
+ * See async_notify_init_callback().
+ */
+#define ASYNC_NOTIFY_METHOD_CALLBACK 3
+
+#define ASYNC_NOTIFY_METHOD_MASK 0x03
+#define ASYNC_NOTIFY_METHOD_POS 0
+
+/**
+ * @brief Identify the region of async_notify flags available for
+ * containing services.
+ *
+ * Bits of the flags field of the async_notify structure at and above
+ * this position may be used by extensions to the async_notify
+ * structure.
+ *
+ * These bits are intended for use by containing service
+ * implementations to record client-specific information.  The bits
+ * are cleared by async_notify_validate().  Use of these does not
```

Comment:
`async_notify_validate` does not seem to clear these bits, they are just ignored there.